panther_analysis_tool/log_schemas/user_defined.py

```diff
@@ -0,0 +1,279 @@
+        return dict(
+            FunctionName=self._LAMBDA_NAME,
+            InvocationType="RequestResponse",
+            Payload=json.dumps({endpoint: payload})
+        )
+
+
+@dataclass
+class UploaderResult:
+    filename: str
+    name: Optional[str]
+    api_response: Optional[Dict[str, Any]] = None
+    definition: Optional[Dict[str, Any]] = None
+    error: Optional[str] = None
+    existed: Optional[bool] = None
+
+
+@dataclass
+class ProcessedFile:
+    error: Optional[str] = None
+    raw: str = ''
+    yaml: Optional[Dict[str, Any]] = None
+
+
+class Uploader:
+    _SCHEMA_NAME_PREFIX = 'Custom.'
+
+    def __init__(self, path: str):
+        self._path = path
+        self._files: Optional[List[str]] = None
+        self._api_client: Optional[Client] = None
+        self._existing_schemas: Optional[List[Dict[str, Any]]] = None
+
+    @property
+    def api_client(self) -> Client:
+        if self._api_client is None:
+            self._api_client = Client()
+        return self._api_client
+
+    @property
+    def files(self) -> List[str]:
+        if self._files is None:
+            self._files = discover_files(self._path)
+        return self._files
+
+    @property
+    def existing_schemas(self) -> List[Dict[str, Any]]:
+        if self._existing_schemas is None:
+            success, response = self.api_client.list_schemas()
+            if not success:
+                raise RuntimeError('unable to retrieve custom schemas')
+            self._existing_schemas = response['results']
+        return self._existing_schemas
+
+    def find_schema(self, name: str) -> Optional[Dict[str, Any]]:
+        for schema in self.existing_schemas:
+            if schema['name'] == name:
+                return schema
+        return None
+
+    def process(self) -> List[UploaderResult]:
+        if not self.files:
+            return []
+
+        processed_files = self._load_from_yaml(self.files)
+        results = []
+        for filename, processed_file in processed_files.items():
+            if processed_file.error is not None:
+                results.append(
+                    UploaderResult(
+                        name=None,
+                        filename=filename,
+                        error=processed_file.error,
+                    )
+                )
+
+        for filename, processed_file in processed_files.items():
+            if processed_file.error is not None:
+                continue
+
+            name, error = self._extract_schema_name(processed_file.yaml)
+            result = UploaderResult(filename=filename, name=name)
+
+            result.name = name
+            result.error = error
+            if not result.error:
+                existed, success, response = self._update_or_create_schema(name, processed_file)
+                result.existed = existed
+                if not success:
+                    api_error = response.get('error')
+                    if api_error is not None:
```

Comment:
What if it is `None`? Should we set the `result.error` to some default value?

Reply:
We enter this path only if `result.error` is not present, so we don't need to reset it, but only keep record of the API error if there's one.